### PR TITLE
PKCS12 support

### PR DESCRIPTION
**Pull request by timgerstel**
This pull request implements the following feature:
- Add support for certificates and certificate authorities stored within PKCS12 files.

Signed-off-by: Timothy Gerstel <tgerstel@rocketsoftware.com>

### CODE REVIEW

File: lib/util.js [hunk 1]
```diff
@@ -151,7 +152,24 @@ module.exports.readFilesToArray = function(fileList) {
   var contentArray = [];
   fileList.forEach(function(filePath) {
     try {
-      contentArray.push(fs.readFileSync(filePath));
+      let extension = filePath.split('.').pop();
+      let content = fs.readFileSync(filePath);
+      if(extension == 'p12'){
```
Comment (lchudinov) [inline lib/util.js:157]: `pfx` extension is a synonym for `p12`. Should the condition take into account both extensions?

Comment (timgerstel) [reply lib/util.js]: Good catch!

File: lib/util.js [hunk 2]
```diff
@@ -151,7 +152,24 @@ module.exports.readFilesToArray = function(fileList) {
...
+        let p12Asn1 = forge.asn1.fromDer(p12Der);
+        let p12;
+        try{
+          p12 = forge.pkcs12.pkcs12FromAsn1(p12Asn1, "");
+        } catch (e) {
+          loggers.bootstrapLogger.warn("ZWED0173W", e.message);
+          p12 = forge.pkcs12.pkcs12FromAsn1(p12Asn1, process.env.KEYSTORE_PASSWORD || "");
+        }
+        let certData = p12.getBags({bagType: forge.pki.oids.certBag});
+        let certBag = certData[forge.pki.oids.certBag][0];
```
Comment (lchudinov) [inline lib/util.js:168]: Why should we read only the first certificate in the chain and ignore the remaining?

**Comment by lchudinov** [comment]
@1000TurquoisePogs app-server already supports the `pfx` format. Please clarify, does this PR introduce new functionality?

**Review by cleonty — APPROVED** [review]
Looks good to me now.

**Review by lchudinov — APPROVED** [review]
Looks good to me now

File: lib/util.js [hunk 3]
```diff
@@ -151,7 +152,26 @@ module.exports.readFilesToArray = function(fileList) {
...
+        let p12Der = forge.util.decode64(content.toString('base64'));
+        let p12Asn1 = forge.asn1.fromDer(p12Der);
+        let p12;
+        try{
+          p12 = forge.pkcs12.pkcs12FromAsn1(p12Asn1, process.env.KEYSTORE_PASSWORD || "");
+        } catch (e) {
+          loggers.bootstrapLogger.warn("ZWED0173W", e.message);
+          p12 = forge.pkcs12.pkcs12FromAsn1(p12Asn1, "");
+        }
+        let certData = p12.getBags({bagType: forge.pki.oids.certBag});
```
Comment (1000TurquoisePogs) [inline lib/util.js:167]: p12 files can be like an archive of multiple types of security objects (keys, certificates, crls etc)
the function this is in is generic: the output is assigned to either keys, certs, cert authorities, depending on which section of the processing we are in.

but, this new code appears to assume its only being used for certs, which is not true.

File: lib/util.js [hunk 4]
```diff
@@ -151,7 +152,26 @@ module.exports.readFilesToArray = function(fileList) {
...
+        for(let i = 0; i < certData.length; i++){
+          let currentCert = certData[forge.pki.oids.certBag][i];
+          contentArray.push(Buffer.from(forge.pki.certificateToPem(currentCert.cert), 'utf8'));
```
Comment (1000TurquoisePogs) [inline lib/util.js:170]: cert is fine when its a cert, but what if its a key? and, how are certificate authorities different from plain certs? same thing? i'm not 100% sure on that.

**Pull request merged**